File: web-app/src/services/authService.js
import api from "./api";

const authService = {
  login: async (email, password) => {
    try {
      const response = await api.post("/auth/login", { email, password });
      return {
        success: true,
        token: response.data.token,
        refreshToken: response.data.refreshToken,
        user: response.data.user,
      };
    } catch (error) {
      console.error("Login error:", error);
      return {
        success: false,
        message: error.response?.data?.message || "Login failed",
      };
    }
  },

  signup: async (userData) => {
    try {
      const response = await api.post("/auth/register", userData);
      return {
        success: true,
        message: "Account created successfully",
        user: response.data.user,
      };
    } catch (error) {
      console.error("Signup error:", error);
      return {
        success: false,
        message: error.response?.data?.message || "Signup failed",
      };
    }
  },

  logout: async () => {
    try {
      await api.post("/auth/logout");
      localStorage.removeItem("token");
      localStorage.removeItem("refreshToken");
      return { success: true };
    } catch (error) {
      console.error("Logout error:", error);
      // Clear tokens even if logout fails
      localStorage.removeItem("token");
      localStorage.removeItem("refreshToken");
      return { success: true };
    }
  },

  verifyToken: async () => {
    try {
      const token = localStorage.getItem("token");
      if (!token) {
        return { success: false, message: "No token found" };
      }

      const response = await api.get("/auth/verify", {
        headers: { Authorization: `Bearer ${token}` },
      });

      return {
        success: true,
        user: response.data.user,
      };
    } catch (error) {
      console.error("Token verification error:", error);
      return {
        success: false,
        message: error.response?.data?.message || "Token verification failed",
      };
    }
  },

  refreshToken: async (refreshToken) => {
    try {
      const response = await api.post("/auth/refresh-token", { refreshToken });
      return {
        success: true,
        token: response.data.token,
        refreshToken: response.data.refreshToken,
      };
    } catch (error) {
      console.error("Token refresh error:", error);
      return {
        success: false,
        message: error.response?.data?.message || "Token refresh failed",
      };
    }
  },

  forgotPassword: async (email) => {
    try {
      const response = await api.post("/auth/forgot-password", { email });
      return {
        success: true,
        message: "Password reset email sent",
      };
    } catch (error) {
      console.error("Forgot password error:", error);
      return {
        success: false,
        message: error.response?.data?.message || "Failed to send reset email",
      };
    }
  },

  resetPassword: async (token, newPassword) => {
    try {
      const response = await api.post("/auth/reset-password", {
        token,
        newPassword,
      });
      return {
        success: true,
        message: "Password reset successfully",
      };
    } catch (error) {
      console.error("Reset password error:", error);
      return {
        success: false,
        message: error.response?.data?.message || "Failed to reset password",
      };
    }
  },

  updateProfile: async (userData) => {
    try {
      const response = await api.put("/auth/profile", userData);
      return {
        success: true,
        user: response.data.user,
        message: "Profile updated successfully",
      };
    } catch (error) {
      console.error("Update profile error:", error);
      return {
        success: false,
        message: error.response?.data?.message || "Failed to update profile",
      };
    }
  },
};

export { authService };

File: IBCM-ios/IBCM/IBCM/Services/AuthService.swift
import Foundation
import Combine

enum AuthState {
    case unknown
    case authenticated
    case notAuthenticated
}

struct AuthResponse {
    let isAuthenticated: Bool
    let user: User?
    let hasSelectedCategories: Bool
    let token: String?
}

class AuthService {
    static let shared = AuthService()
    
    // API endpoints
    private let loginEndpoint = "auth/login"
    private let signupEndpoint = "auth/register"
    private let forgotPasswordEndpoint = "auth/forgot-password"
    private let resetPasswordEndpoint = "auth/reset-password"
    private let verifyResetTokenEndpoint = "auth/validate-reset-token"
    private let logoutEndpoint = "auth/logout"
    private let checkAuthEndpoint = "auth/check"
    
    // API service for network calls
    private let apiService = APIService.shared
    
    // Publisher for auth state changes
    private let authStateSubject = CurrentValueSubject<AuthState, Never>(.unknown)
    var authStatePublisher: AnyPublisher<AuthState, Never> {
        authStateSubject.eraseToAnyPublisher()
    }
    
    // MARK: - Auth Methods
    
    func login(email: String, password: String) async throws -> AuthResponse {
        let loginRequest = LoginRequest(email: email, password: password)
        let data = try JSONEncoder().encode(loginRequest)
        
        let response: AuthData = try await apiService.request(
            endpoint: loginEndpoint,
            method: "POST",
            body: data
        )
        
        // Save token
        apiService.setAuthToken(response.token)
        
        // Update auth state
        authStateSubject.send(.authenticated)
        
        return AuthResponse(
            isAuthenticated: true,
            user: response.user,
            hasSelectedCategories: response.user.interests.count > 0,
            token: response.token
        )
    }
    
    func signup(name: String, email: String, password: String) async throws -> AuthResponse {
        let names = name.split(separator: " ")
        let firstName = String(names.first ?? "")
        let lastName = names.count > 1 ? names.dropFirst().joined(separator: " ") : ""
        
        let registerRequest = RegisterRequest(
            firstName: firstName,
            lastName: lastName,
            email: email,
            password: password,
            phoneNumber: nil,
            dateOfBirth: nil,
            gender: nil,
            location: nil,
            interests: nil,
            marketingConsent: false,
            termsAccepted: true
        )
        
        let data = try JSONEncoder().encode(registerRequest)
        
        let response: AuthData = try await apiService.request(
            endpoint: signupEndpoint,
            method: "POST",
            body: data
        )
        
        // Save token
        apiService.setAuthToken(response.token)
        
        // Update auth state
        authStateSubject.send(.authenticated)
        
        return AuthResponse(
                        isAuthenticated: true,
            user: response.user,
            hasSelectedCategories: response.user.interests.count > 0,
            token: response.token
        )
    }
    
    func forgotPassword(email: String) async throws -> Bool {
        let forgotPasswordRequest = ForgotPasswordRequest(email: email)
        let data = try JSONEncoder().encode(forgotPasswordRequest)
        
        let response: ApiResponse<Bool> = try await apiService.request(
            endpoint: forgotPasswordEndpoint,
            method: "POST",
            body: data
        )
        
        return response.data ?? false
    }
    
    func resetPassword(token: String, newPassword: String) async throws -> Bool {
        let resetPasswordRequest = ResetPasswordRequest(
            token: token,
            newPassword: newPassword,
            confirmPassword: newPassword
        )
        
        let data = try JSONEncoder().encode(resetPasswordRequest)
        
        let response: ApiResponse<Bool> = try await apiService.request(
            endpoint: "\(resetPasswordEndpoint)/\(token)",
            method: "POST",
            body: data
        )
        
        return response.data ?? false
    }
    
    func verifyResetToken(token: String) async throws -> Bool {
        let response: ApiResponse<Bool> = try await apiService.request(
            endpoint: "\(verifyResetTokenEndpoint)/\(token)",
            method: "GET"
        )
        
        return response.data ?? false
    }
    
    func logout() async throws {
        let _: ApiResponse<Bool> = try await apiService.request(
            endpoint: logoutEndpoint,
            method: "POST"
        )
        
        // Clear token
        apiService.setAuthToken(nil)
        
        // Update auth state
        authStateSubject.send(.notAuthenticated)
    }
    
    func checkAuthState() async throws -> AuthResponse {
        do {
            let response: AuthData = try await apiService.request(
                endpoint: checkAuthEndpoint,
                method: "GET"
            )
            
            // Save token if present
            if let token = response.token {
                apiService.setAuthToken(token)
            }
            
            // Update auth state
            authStateSubject.send(.authenticated)
            
            return AuthResponse(
                        isAuthenticated: true,
                user: response.user,
                hasSelectedCategories: response.user.interests.count > 0,
                token: response.token
            )
        } catch {
            // Clear token
            apiService.setAuthToken(nil)
            
            // Update auth state
            authStateSubject.send(.notAuthenticated)
            
            return AuthResponse(
                        isAuthenticated: false,
                        user: nil,
                hasSelectedCategories: false,
                token: nil
            )
        }
    }
}

// MARK: - API Response Models
struct ApiResponse<T: Codable>: Codable {
    let success: Bool
    let data: T?
    let message: String?
    let errors: [String]?
}
